File: src/main/java/ca/odell/glazedlists/impl/GlazedListsImpl.java
/* Glazed Lists                                                 (c) 2003-2006 */
/* http://publicobject.com/glazedlists/                      publicobject.com,*/
/*                                                     O'Dell Engineering Ltd.*/
package ca.odell.glazedlists.impl;

import ca.odell.glazedlists.EventList;
import ca.odell.glazedlists.GlazedLists;

import java.util.*;
import java.util.function.Function;

/**
 * A utility class containing all sorts of random things that are useful for
 * implementing Glazed Lists but not for using Glazed Lists.
 *
 * @author <a href="mailto:jesse@swank.ca">Jesse Wilson</a>
 */
public final class GlazedListsImpl {

    /**
     * A dummy constructor to prevent instantiation of this class
     */
    private GlazedListsImpl() {
        throw new UnsupportedOperationException();
    }

    // Utility Methods // // // // // // // // // // // // // // // // // // //

    /**
     * Replace all elements in the target {@link EventList} with the elements in
     * the source {@link List}.
     *
     * <p>Note that both collections must be sorted prior to execution using the
     * {@link Comparator} specified.
     */
    public static <E> void replaceAll(EventList<E> target, Collection<E> source, boolean updates, Comparator<E> comparator) {
        // use the default comparator if none is specified
        if(comparator == null) {
            comparator = (Comparator<E>)GlazedLists.comparableComparator();
        }

        // walk through each list simultaneously
        int targetIndex = -1;
        Iterator<E> sourceIterator = source.iterator();

        // define marker objects that mean we need new objects
        final E NEW_VALUE_NEEDED = (E)Void.class;
        E targetObject = NEW_VALUE_NEEDED;
        E sourceObject = NEW_VALUE_NEEDED;

        // while there are elements left in either list
        while(true) {

            // do our best to get new objects if necessary
            if(targetObject == NEW_VALUE_NEEDED) {
                if(targetIndex < target.size()) targetIndex++;
                if(targetIndex < target.size()) targetObject = target.get(targetIndex);
            }
            if(sourceObject == NEW_VALUE_NEEDED) {
                if(sourceIterator.hasNext()) sourceObject = sourceIterator.next();
            }

            // we've exhausted our dataset
            if(targetObject == NEW_VALUE_NEEDED && sourceObject == NEW_VALUE_NEEDED) {
                break;
            }

            // figure out if this is an insert, update or delete on the target
            int compareResult;
            if(targetObject == NEW_VALUE_NEEDED) compareResult = 1;
            else if(sourceObject == NEW_VALUE_NEEDED) compareResult = -1;
            else compareResult = comparator.compare(targetObject, sourceObject);

            // the target value precedes the source value, delete it
            if(compareResult < 0) {
                target.remove(targetIndex);
                targetIndex--;
                targetObject = NEW_VALUE_NEEDED;

            // the values are equal, keep them
            } else if(compareResult == 0) {
                if(updates) {
                    target.set(targetIndex, sourceObject);
                }
                targetObject = NEW_VALUE_NEEDED;
                sourceObject = NEW_VALUE_NEEDED;

            // the source value precedes the target, insert it
            } else if(compareResult > 0) {
                target.add(targetIndex, sourceObject);
                targetIndex++;
                sourceObject = NEW_VALUE_NEEDED;
            }
        }
    }

    /**
     * Returns a non-symmetric Comparator that returns 0 if two Objects are
     * equal as specified by {@link Object#equals(Object)}, or 1 otherwise.
     */
    public static <T> Comparator<T> equalsComparator() {
        return new EqualsComparator<>();
    }
    private static class EqualsComparator<T> implements Comparator<T> {
        @Override
        @SuppressWarnings("ComparatorMethodParameterNotUsed")
        public int compare(T alpha, T beta) {
            return Objects.equals(alpha, beta) ? 0 : 1;
        }
    }

    /**
     * Returns a {@link Function} that simply reflects the
     * function's argument as its result.
     */
    public static <E> Function<E,E> identityFunction() {
        return Function.identity();
    }
}

File: src/main/kotlin/ca/odell/glazedlists/impl/GlazedListsImpl.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package ca.odell.glazedlists.impl

import ca.odell.glazedlists.EventList
import ca.odell.glazedlists.GlazedLists
import java.util.function.Function

/** Internal utilities used by Glazed Lists implementations. */
@Suppress("UNCHECKED_CAST")
object GlazedListsImpl {
    @JvmStatic
    fun <E> replaceAll(
        target: EventList<E>,
        source: Collection<@JvmSuppressWildcards E>,
        updates: Boolean,
        comparator: Comparator<E>?,
    ) {
        val actualComparator = comparator
            ?: GlazedLists.comparableComparator<Comparable<Any?>>() as Comparator<E>
        val newValueNeeded = Any()
        val sourceIterator = source.iterator()
        var targetIndex = -1
        var targetObject: Any? = newValueNeeded
        var sourceObject: Any? = newValueNeeded

        while (true) {
            if (targetObject === newValueNeeded) {
                if (targetIndex < target.size) targetIndex++
                if (targetIndex < target.size) targetObject = target[targetIndex]
            }
            if (sourceObject === newValueNeeded && sourceIterator.hasNext()) {
                sourceObject = sourceIterator.next()
            }

            if (targetObject === newValueNeeded && sourceObject === newValueNeeded) break

            val compareResult = when {
                targetObject === newValueNeeded -> 1
                sourceObject === newValueNeeded -> -1
                else -> actualComparator.compare(targetObject as E, sourceObject as E)
            }

            when {
                compareResult < 0 -> {
                    target.removeAt(targetIndex)
                    targetIndex--
                    targetObject = newValueNeeded
                }

                compareResult == 0 -> {
                    if (updates) target[targetIndex] = sourceObject as E
                    targetObject = newValueNeeded
                    sourceObject = newValueNeeded
                }

                else -> {
                    target.add(targetIndex, sourceObject as E)
                    targetIndex++
                    sourceObject = newValueNeeded
                }
            }
        }
    }

    @JvmStatic
    fun <T> equalsComparator(): Comparator<T> = EqualsComparator()

    @JvmStatic
    fun <E> identityFunction(): Function<E, E> = Function.identity()

    private class EqualsComparator<T> : Comparator<T> {
        override fun compare(alpha: T, beta: T): Int = if (alpha == beta) 0 else 1
    }
}
